File: app/src/main/java/com/harvey/arcfacedamo/ui/FaceRegisterActivity.java
package com.harvey.arcfacedamo.ui;

import android.app.AlertDialog;
import android.content.pm.ActivityInfo;
import android.graphics.Bitmap;
import android.hardware.Camera;
import android.os.Build;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.RadioGroup;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.arcsoft.face.FaceInfo;
import com.guo.android_extend.widget.ExtImageView;
import com.harvey.arcface.AIFace;
import com.harvey.arcface.model.CameraModel;
import com.harvey.arcface.model.FaceAction;
import com.harvey.arcface.model.FaceCameraModel;
import com.harvey.arcface.utils.FaceUtils;
import com.harvey.arcface.view.SurfaceViewCamera;
import com.harvey.arcface.view.SurfaceViewSaveFace;
import com.harvey.arcfacedamo.R;
import com.harvey.arcfacedamo.utils.FaceMatchHelper;
import com.harvey.arcfacedamo.utils.ToastUtil;

/**
 * Created by harvey on 2018/1/12.
 */

public class FaceRegisterActivity extends AppCompatActivity
        implements
        Camera.PreviewCallback {
    SurfaceViewSaveFace surfaceViewSaveFace;
    SurfaceViewCamera surfaceViewCamera;
    View dialogLayout;
    ExtImageView dialogExtImageView;
    EditText dialogName;
    EditText dialogAge;
    RadioGroup dialogSex;
    String faceName;
    String faceAge;
    String faceSex;
    AlertDialog registerDialog;
    AIFace mAiFace;
    FaceMatchHelper faceRegisterHelper;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getContentViewId());
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            WindowManager.LayoutParams attributes = getWindow().getAttributes();
            attributes.systemUiVisibility = View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION;
            getWindow().setAttributes(attributes);
        }
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LOCKED);
        initHolder();
        initListener();
        initData();
    }

    protected int getContentViewId() {
        return R.layout.activity_register_face;
    }

    protected void initHolder() {
        surfaceViewCamera = findViewById(R.id.surfaceView_Camera);
        surfaceViewSaveFace = findViewById(R.id.surfaceViewSaveFace);
        dialogLayout = LayoutInflater.from(this).inflate(R.layout.dialog_save_face, null);
        dialogExtImageView = dialogLayout.findViewById(R.id.extimageview);
        dialogName = dialogLayout.findViewById(R.id.et_name);
        dialogAge = dialogLayout.findViewById(R.id.et_age);
        dialogSex = dialogLayout.findViewById(R.id.rg_sex);

    }

    public void initListener() {
        dialogSex.setOnCheckedChangeListener((group, checkedId) -> {
            if (checkedId == R.id.rb_man) {
                faceSex = "男";
            }
            if (checkedId == R.id.rb_women) {
                faceSex = "女";
            }
        });
    }

    protected void initData() {
        mAiFace = new AIFace
                .Builder(this)
                .combinedMask(FaceAction.DETECT_FACE_FEATURE)
                .orientPriority(surfaceViewCamera.getCameraDisplayOrientation())
                .build();
        faceRegisterHelper = new FaceMatchHelper(this, mAiFace);
        AIFace.showLog(true);
        surfaceViewCamera.setCameraCallBack(this);
        surfaceViewSaveFace.uploadTimeSecondDown(1);
        surfaceViewSaveFace.setSaveFaceListener(new SurfaceViewSaveFace.SaveFaceListener() {
            @Override
            public void onSuccess(FaceCameraModel faceModel) {
                if (faceModel != null && faceModel.getFaceInfo().size() > 0) {
                    showSaveFaceDialog(faceModel.getFaceInfo().get(0), faceModel);
                } else {
                    surfaceViewSaveFace.reset();
                }
            }

            @Override
            public void onTimeSecondDown(int TimeSecond) {
                Log.e("harvey", "onTimeSecondDown---->" + TimeSecond);
            }

            @Override
            public void onErrorMsg(int errorCode) {
                Log.e("harvey", "onErrorMsg---->" + errorCode);
            }
        });
        surfaceViewSaveFace.setDisplayOrientation(surfaceViewCamera.getCameraDisplayOrientation());
        surfaceViewSaveFace.setFrontCamera(surfaceViewCamera.isFront());

    }

    public void showSaveFaceDialog(final FaceInfo faceInfo, final CameraModel cameraModel) {
        Log.e("harvey", String.format("showSaveFaceDialog-->FaceInfo:%d,Nv21:%s", faceInfo.hashCode(), cameraModel.getNv21()));
        if (registerDialog != null && registerDialog.isShowing()) {
            registerDialog.dismiss();
        }
        ViewParent parent = dialogLayout.getParent();
        if (parent != null) {
            ((ViewGroup) parent).removeAllViews();
        }
        final Bitmap faceBitmap = FaceUtils.getFaceBitmap(faceInfo, cameraModel);
        dialogExtImageView.setImageBitmap(faceBitmap);
        registerDialog = new AlertDialog.Builder(this).
                setTitle("是否注册该图片?")
                .setIcon(android.R.drawable.ic_dialog_info)
                .setView(dialogLayout).setPositiveButton("确定", null)
                .setNegativeButton("取消", (dialog, which) -> {
                    if (!faceBitmap.isRecycled())
                        faceBitmap.recycle();
                    surfaceViewSaveFace.reset();
                    dialog.dismiss();
                }).create();
        registerDialog.show();
        registerDialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                faceName = dialogName.getText().toString().trim();
                faceAge = dialogAge.getText().toString().trim();
                if (TextUtils.isEmpty(faceName)) {
                    ToastUtil.showToast(FaceRegisterActivity.this, "请输入姓名！");
                } else if (TextUtils.isEmpty(faceAge)) {
                    ToastUtil.showToast(FaceRegisterActivity.this, "请输入年龄！");
                } else if (TextUtils.isEmpty(faceSex)) {
                    ToastUtil.showToast(FaceRegisterActivity.this, "请选择性别！");
                } else {
                    boolean result = faceRegisterHelper.registerNv21(faceInfo, cameraModel, faceName,
                            Integer.valueOf(faceAge), faceSex, getApplication().getExternalCacheDir().getPath());
                    if (result)
                        ToastUtil.showToast(FaceRegisterActivity.this, "注册人脸成功！");
                    else
                        ToastUtil.showToast(FaceRegisterActivity.this, "注册人脸失败！");
                    if (!faceBitmap.isRecycled())
                        faceBitmap.recycle();
                    surfaceViewSaveFace.reset();
                    registerDialog.dismiss();
                }
            }
        });
        registerDialog.setCanceledOnTouchOutside(false);
    }

    public void switchCamera(View view) {
        surfaceViewCamera.switchCamera();
        surfaceViewSaveFace.setFrontCamera(surfaceViewCamera.isFront());
    }

    @Override
    public void onPreviewFrame(byte[] data, Camera camera) {
        Camera.Size size = camera.getParameters().getPreviewSize();
        FaceCameraModel faceCameraModel = mAiFace.detectFaceWithCamera(data, size.width, size.height);
        surfaceViewSaveFace.uploadFace(faceCameraModel);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mAiFace.destroy();
    }
}

File: kotlin-app/src/main/java/com/harvey/arcfacedemo/ui/FaceRegisterActivity.kt
package com.harvey.arcfacedemo.ui

import android.app.AlertDialog
import android.content.Context
import android.hardware.Camera
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.harvey.arcface.ArcFaceEngine
import com.harvey.arcface.DetectFaceAction
import com.harvey.arcface.model.FaceFindCameraModel
import com.harvey.arcface.model.FaceFindModel
import com.harvey.arcface.view.SurfaceViewSaveFace
import com.harvey.arcfacedemo.R
import kotlinx.android.synthetic.main.activity_register_face.*
import kotlinx.android.synthetic.main.dialog_save_face.view.*

/**
 * Created by hanhui on 2018/6/1 0001 13:53
 */
class FaceRegisterActivity : AppCompatActivity(), Camera.PreviewCallback, DetectFaceAction.OnFaceDetectListener {
    private val mDetectFaceAction: DetectFaceAction by lazy {
        DetectFaceAction()
    }
    private lateinit var dialogLayout: View
    private var faceName = ""
    private var faceAge = "0"
    private var faceSex = ""
    private var registerDialog: AlertDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register_face)
        initHolder()
        initListener()
        initData()
    }

    private fun initHolder() {
        dialogLayout = LayoutInflater.from(this).inflate(R.layout.dialog_save_face, null)
    }

    private fun initListener() {
        dialogLayout.rg_sex.setOnCheckedChangeListener { _, checkedId ->
            if (checkedId == R.id.rb_man) {
                faceSex = "男"

            }
            if (checkedId == R.id.rb_women) {
                faceSex = "女"
            }
        }
    }

    private fun initData() {
        mDetectFaceAction.setOnFaceDetectListener(this)
        surfaceViewCamera.setCameraCallBack(this)
        surfaceViewSaveFace.uploadTimeSecondDown(1)
        surfaceViewSaveFace.setSaveFaceListener(object : SurfaceViewSaveFace.SaveFaceListener {

            override fun onSuccess(faceModel: FaceFindCameraModel) {
                if (faceModel.faceFindModels.isNotEmpty()) {
                    showSaveFaceDialog(faceModel.faceFindModels[0], faceModel.cameraData)
                }
            }

            override fun onTimeSecondDown(TimeSecond: Int) {
                Log.e("harvey", "onTimeSecondDown---->" + TimeSecond)
            }

            override fun onErrorMsg(errorCode: Int) {
                Log.e("harvey", "onErrorMsg---->" + errorCode)
            }
        })
    }

    fun showSaveFaceDialog(faceModel: FaceFindModel, data: ByteArray) {
        if (registerDialog != null && registerDialog!!.isShowing) {
            registerDialog?.dismiss()
        }
        val parent = dialogLayout.parent
        if (parent != null) {
            (parent as ViewGroup).removeAllViews()
        }
        val faceBitmap = ArcFaceEngine.getFaceBitmap(faceModel, data)
        dialogLayout.extimageview.setImageBitmap(faceBitmap)
        registerDialog = AlertDialog.Builder(this).setTitle("是否注册该图片?").setIcon(android.R.drawable.ic_dialog_info)
                .setView(dialogLayout).setPositiveButton("确定", null)
                .setNegativeButton("取消") { dialog, which ->
                    if (!faceBitmap.isRecycled)
                        faceBitmap.recycle()
                    surfaceViewSaveFace.reset()
                    dialog.dismiss()
                }.create()
        registerDialog?.show()
        registerDialog!!.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
            faceName = dialogLayout.et_name.text.toString().trim { it <= ' ' }
            faceAge = dialogLayout.et_age.text.toString().trim { it <= ' ' }
            when {
                TextUtils.isEmpty(faceName) -> showToast(this, "请输入姓名！")
                TextUtils.isEmpty(faceAge) -> showToast(this, "请输入年龄！")
                TextUtils.isEmpty(faceSex) -> showToast(this, "请选择性别！")
                else -> {
                    val result = ArcFaceEngine.saveFace(data, faceModel, faceName,
                            faceAge.toInt(), faceSex, application.externalCacheDir.path)
                    if (result)
                        showToast(this, "注册人脸成功！")
                    else
                        showToast(this, "注册人脸失败！")
                    if (!faceBitmap.isRecycled)
                        faceBitmap.recycle()
                    surfaceViewSaveFace.reset()
                    registerDialog!!.dismiss()
                }
            }
        }
        registerDialog?.setCanceledOnTouchOutside(false)
    }

    private fun showToast(context: Context, msg: String) {
        Toast.makeText(context.applicationContext, msg, Toast.LENGTH_SHORT).show()

    }

    fun switchCamera(view: View) {
        surfaceViewCamera.switchCamera()
    }

    override fun onPreviewFrame(data: ByteArray, camera: Camera) {
        val size = camera.parameters.previewSize
        mDetectFaceAction.detectFace(data, size.width, size.height)
    }

    override fun onFaceDetect(faceFindModels: List<FaceFindModel>, frameBytes: ByteArray) {
        surfaceViewSaveFace.uploadFace(faceFindModels, frameBytes)
    }

    override fun onDestroy() {
        super.onDestroy()
        mDetectFaceAction.destroy()
    }
}
